Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>

pbelevich opened the pull request:
This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35167 Move normal() to DistributionTemplates
* #34774 pybind11 type_caster for at::Generator and custom RNG python test
* **#34468 Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>**

Differential Revision: [D20549420](https://our.internmc.facebook.com/intern/diff/D20549420)

pbelevich committed:
GeneratorHolder

[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] GeneratorHolder"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. Move `at::Generator` to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. Recreate `at::Generator` as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Replace most of `at::Generator*` with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue)

* **#34468 [WIP] GeneratorHolder**

[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

ezyang commented on aten/src/ATen/CPUGenerator.cpp:
```diff
@@ -14,16 +14,16 @@ namespace detail {
-  return &default_gen_cpu;
+Generator getDefaultCPUGenerator() {
+  static auto default_gen_cpu = createCPUGenerator(c10::detail::getNonDeterministicRandom());
+  return default_gen_cpu;
```
Not really sure if you should be returning `Generator` here or `const Generator&` here. Latter might be better as this generator always stays live and there's no reason to force a refcount bump here.

pbelevich replied:
addressed

ezyang commented on aten/src/ATen/CPUGenerator.h:
```diff
@@ -3,10 +3,11 @@
 namespace at {
 
-struct CAFFE2_API CPUGenerator : public Generator {
+struct CAFFE2_API CPUGenerator : public c10::GeneratorImpl {
```
don't have to do it in this PR, but CPUGenerator and friends should probably also get the Impl suffix too

pbelevich replied:
#34984

ezyang commented on aten/src/ATen/Context.h:
```diff
@@ -214,22 +214,22 @@ static inline bool hasMKLDNN() {
     // See Note [Acquire lock when using random generators]
-    std::lock_guard<std::mutex> lock(gen.mutex_);
-    gen.set_current_seed(seed);
+    std::lock_guard<std::mutex> lock(gen->mutex_);
```
Also doesn't have to be this PR, but we should consider hiding the mutex behind a method instead of exposing all the field guts to everyone.

pbelevich replied:
#34987

ezyang commented on aten/src/ATen/Utils.h:
```diff
@@ -139,32 +139,30 @@ inline int64_t prod_intlist(ArrayRef<int64_t> list) {
-static inline T * check_generator(Generator * expr) {
-  if (T::device_type() == expr->device().type()) {
-    return static_cast<T*>(expr);
+static inline T* get_generator_or_default(Generator expr, Generator defaultValue) {
```
These can take the generator by reference, the function doesn't take ownership

pbelevich replied:
addressed

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -49,45 +50,56 @@
-  std::shared_ptr<Generator> clone() const;
+  // TODO(pbelevich): delete this after replace Generator generator = nullptr with {}
+  Generator(c10::GeneratorImpl* gen_impl)
+   : impl_(std::shared_ptr<c10::GeneratorImpl>(gen_impl)) {}
```
A much safer variant of this overload is `Generator(nullptr_t gen_impl)` which will only permit `nullptr` exactly.

ezyang replied:
(If I read the TODO correctly, that is!)

pbelevich replied:
addressed

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -49,45 +50,56 @@
+  Generator& operator=(Generator&& x) & {
+    impl_ = std::move(x.impl_);
+    return *this;
+  }
```
You shouldn't need to define these explicitly; the default implementations should do the right thing.

pbelevich replied:
removed

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -49,45 +50,56 @@
+  c10::GeneratorImpl* operator->() const { return impl_.get(); }
 
-CAFFE2_API uint64_t getNonDeterministicRandom(bool is_cuda = false);
+  c10::GeneratorImpl* get() const { return impl_.get(); }
```
This is rarely what you actually want, right, because the abstract `GeneratorImpl` class generally doesn't have enough information for you to do anything useful with it. Instead, I'd suggest axing this method, and make the typed cast to CPUGenerator/CUDAGenerator template function a method on Generator itself.

pbelevich replied:
addressed

ezyang commented on aten/src/ATen/cuda/CUDAGenerator.cpp:
```diff
@@ -44,25 +45,25 @@ CUDAGenerator* getDefaultCUDAGenerator(DeviceIndex device_index) {
-  auto gen = std::make_shared<CUDAGenerator>(idx);
-  gen->set_current_seed(default_rng_seed_val);
-  gen->set_philox_offset_per_thread(0);
+  auto gen = make_generator<CUDAGenerator>(idx);
```
With my C++ API hat on, we should endeavor to make the API for allocating a CUDA generator look as much as possible as the Python API. (So, not `make_generator`, in general! `make_generator` still a useful helper function for internal implementation reasons though, I guess.)

pbelevich replied:
possible solution is adding `Generator(Device device)`, please see the next PR

ezyang replied:
Constructor with device is probably not the way we should do it :) We'll talk about it on the other PR.

ezyang commented on aten/src/ATen/cuda/CUDAGenerator.cpp:
```diff
@@ -44,25 +45,25 @@ CUDAGenerator* getDefaultCUDAGenerator(DeviceIndex device_index) {
-  gen->set_current_seed(default_rng_seed_val);
-  gen->set_philox_offset_per_thread(0);
+  auto gen = make_generator<CUDAGenerator>(idx);
+  check_generator<CUDAGenerator>(gen)->set_current_seed(default_rng_seed_val);
```
This is kind of goofy. Better to just check generator once and then use the resulting handle to do the mutations you need, right?

pbelevich replied:
addressed

ezyang commented on aten/src/ATen/test/cpu_generator_test.cpp:
```diff
@@ -54,19 +54,19 @@ TEST(CPUGenerator, TestMultithreadingGetEngineOperator) {
   {
     std::lock_guard<std::mutex> lock(gen1->mutex_);
-    gen2 = gen1->clone(); // capture the current state of default generator
+    gen2 = Generator(gen1->clone()); // capture the current state of default generator
```
Why is the `Generator()` constructor call needed here?

pbelevich replied:
Because `clone()` is a method of `at::GeneratorImpl` and returns `std::shared_ptr<at::GeneratorImpl>` and c-tor `at::Generator(std::shared_ptr<at::GeneratorImpl>)` is explicit

ezyang replied:
Maybe there should be a `Generator::clone()` for this case?

pbelevich replied:
added

ezyang commented on aten/src/ATen/test/rng_test.h:
```diff
@@ -72,7 +72,7 @@ void test_random_from_to(const at::Device& device) {
       if (!to.has_value() || from < *to) {
         for (const uint64_t val : vals) {
-          auto gen = new RNG(val);
+          auto gen = at::make_generator<RNG>(val);
```
Music to my ears!!

ezyang commented on torch/csrc/Generator.cpp:
```diff
@@ -24,22 +24,18 @@ using namespace torch;
 {
-  if (self->owner) {
-    delete self->cdata;
-  }
```
Whoops! You're leaking the generators here. Because the PyObject now directly contains a C++ type, you must placement delete it on deallocation.

pbelevich replied:
addressed

ezyang replied:
First, sorry, I misspoke in my comment; it's not placement delete you need to do (placement delete is not a thing), you need to just explicitly call the destructor. Assignment with the default constructor isn't sufficient, as a default constructed Generator may still have a nontrivial destructor.

```
self->cdata->~Generator();
```

ezyang commented on torch/csrc/Generator.h:
```diff
@@ -7,14 +7,14 @@
-  at::Generator *cdata;
-  bool owner;  // if true, frees cdata in destructor
+  at::Generator cdata;
+  // bool owner;  // if true, frees cdata in destructor
```
nit: just delete this

pbelevich replied:
addressed

ezyang reviewed (changes requested):
This is great work, a giant step in the right direction. Unfortunately, there's a memory leak which needs to be fixed before this can go in.

Some broader things to keep in mind (they don't have to be done in this PR):
1. We should consider taking `const Generator&` instead of `Generator` as it avoids a refcount bump. (cc @zdevito for any extra opinions)
2. We should think about how to make it easy for user code to do defaulting behavior correctly for generators. This is kind of entwined with Tensor, which has a default constructor, but for which in our user API we encourage people to accept `optional<Tensor>` in cases when the Tensor may not be specified. In the case of generator, the right call might be to NOT use optional, but we need to make it easy for people to avoid accidentally doing a null pointer dereference. I don't think this should block this PR either, as null pointer was always a problem for the old API too.

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

ezyang commented on aten/src/ATen/Context.h:
```diff
@@ -214,22 +214,22 @@ static inline bool hasMKLDNN() {
-        std::lock_guard<std::mutex> lock(cuda_gen.mutex_);
-        cuda_gen.set_current_seed(seed);
+        std::lock_guard<std::mutex> lock(cuda_gen->mutex_);
+        cuda_gen->set_current_seed(seed);
```
Also doesn't have to be this PR, but we should consider adding methods like `set_current_seed` to Generator directly, so dot-syntax works on them (making Generator consistent with Tensor)

pbelevich replied:
will do in #34987

ezyang commented on aten/src/ATen/Utils.h:
```diff
@@ -139,32 +139,30 @@ inline int64_t prod_intlist(ArrayRef<int64_t> list) {
-    return static_cast<T*>(expr);
+static inline T* get_generator_or_default(const Generator& expr, const Generator& defaultValue) {
+  T* result = expr.defined() ? check_generator<T>(expr) : check_generator<T>(defaultValue);
+  if (result == nullptr) {
```
When can this be nullptr?

pbelevich replied:
I rewrote this part, please take a look

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -49,45 +51,48 @@
-  virtual ~Generator() = default;
-  std::shared_ptr<Generator> clone() const;
+  // TODO(pbelevich): delete this after replace Generator generator = nullptr with c10::optional<at::Generator> = c10::nullopt
+  Generator(std::nullptr_t gen_impl) {}
```
It's pretty important for us to get rid of this TODO soon, as it's pretty confusing that:

```
std::shared_ptr<c10::GeneratorImpl> impl = nullptr;
Generator g(impl); // errors

Generator g(nullptr); // this is OK?!
```

pbelevich replied:
Totally agree, #35067

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -49,45 +51,48 @@
-  virtual uint64_t seed() = 0;
-  Device device() const;
+  bool operator==(const Generator& rhs) const {
+    return (!(this->impl_) && !(rhs.impl_)) || (this->impl_ == rhs.impl_);
```
Can't this just be `this->impl_ == rhs->impl_`? Because if they're both nullptr they'll compare equal.

pbelevich replied:
addressed

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -49,45 +51,48 @@
 
-  DispatchKeySet key_set() const { return key_set_; }
+  bool defined() const {
+    return (bool)impl_;
```
nit: `static_cast<bool>(impl_)` for extra safety. (Or just let implicit conversion on pointer take care of it.)

pbelevich replied:
addressed

ezyang reviewed (approved):
LGTM.

ezyang commented:
Please take care of replacing the `= {}` with explicit destructor call before landing

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)



[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)


Differential Revision: [D20549420](https://our.internmc.facebook.com/intern/diff/D20549420)

[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)


Differential Revision: [D20549420](https://our.internmc.facebook.com/intern/diff/D20549420)

[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)


Differential Revision: [D20549420](https://our.internmc.facebook.com/intern/diff/D20549420)

[ghstack-poisoned]

pbelevich committed:
Update on "Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>"

This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)


Differential Revision: [D20549420](https://our.internmc.facebook.com/intern/diff/D20549420)

[ghstack-poisoned]
